File: app/src/main/java/com/example/frienddebt/dsa/DebtSimplifier.java
package com.example.frienddebt.dsa;

import com.example.frienddebt.model.CashbookEntry;
import com.example.frienddebt.model.DebtEdge;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;

public class DebtSimplifier {

    public static List<DebtEdge> simplifyDebts(List<CashbookEntry> entries) {
        // Step 1: Compute net balances for everyone
        Map<String, Double> balances = new HashMap<>();

        for (CashbookEntry entry : entries) {
            String type = entry.getType();

            if ("SETTLEMENT".equals(type)) {
                // For a settlement, paidBy paid someone else.
                // We store the receiver in the 'participants' list at index 0.
                String payer = entry.getPaidBy();
                String receiver = entry.getParticipants() != null && !entry.getParticipants().isEmpty() ? entry.getParticipants().get(0) : null;
                
                if (payer != null && receiver != null) {
                    double amt = entry.getAmount();
                    balances.put(payer, balances.getOrDefault(payer, 0.0) + amt);
                    balances.put(receiver, balances.getOrDefault(receiver, 0.0) - amt);
                }
            } else if ("EXPENSE".equals(type) && entry.getSplits() != null) {
                // Shared Expense
                String payer = entry.getPaidBy();
                if (payer == null) continue;

                double totalPaid = entry.getAmount();
                balances.put(payer, balances.getOrDefault(payer, 0.0) + totalPaid);

                for (Map.Entry<String, Double> split : entry.getSplits().entrySet()) {
                    String userId = split.getKey();
                    double owe = split.getValue();
                    balances.put(userId, balances.getOrDefault(userId, 0.0) - owe);
                }
            }
        }

        // Step 2: Split into debtors (negative balance) and creditors (positive balance)
        // Max-heaps based on absolute value
        PriorityQueue<Pair> debtors = new PriorityQueue<>((a, b) -> Double.compare(b.amount, a.amount));
        PriorityQueue<Pair> creditors = new PriorityQueue<>((a, b) -> Double.compare(b.amount, a.amount));

        for (Map.Entry<String, Double> b : balances.entrySet()) {
            double amount = Math.round(b.getValue() * 100.0) / 100.0; // Avoid floating point issues
            if (amount < 0) {
                debtors.add(new Pair(b.getKey(), -amount));
            } else if (amount > 0) {
                creditors.add(new Pair(b.getKey(), amount));
            }
        }

        // Step 3: Greedily settle debts
        List<DebtEdge> suggestedPayments = new ArrayList<>();

        while (!debtors.isEmpty() && !creditors.isEmpty()) {
            Pair debtor = debtors.poll();
            Pair creditor = creditors.poll();

            double settleAmount = Math.min(debtor.amount, creditor.amount);
            
            suggestedPayments.add(new DebtEdge(debtor.userId, creditor.userId, settleAmount));

            debtor.amount -= settleAmount;
            creditor.amount -= settleAmount;

            // Push back if there is remaining balance
            if (Math.round(debtor.amount * 100.0) / 100.0 > 0) {
                debtors.add(debtor);
            }
            if (Math.round(creditor.amount * 100.0) / 100.0 > 0) {
                creditors.add(creditor);
            }
        }

        return suggestedPayments;
    }

    private static class Pair {
        String userId;
        double amount;

        Pair(String userId, double amount) {
            this.userId = userId;
            this.amount = amount;
        }
    }
}

File: web/src/lib/DebtSimplifier.ts
export interface DebtEdge {
  from: string;
  to: string;
  amount: number;
}

export function simplifyDebts(transactions: DebtEdge[]): DebtEdge[] {
  // Calculate net balances for each person
  const balances: Record<string, number> = {};

  for (const t of transactions) {
    if (!balances[t.from]) balances[t.from] = 0;
    if (!balances[t.to]) balances[t.to] = 0;

    balances[t.from] -= t.amount;
    balances[t.to] += t.amount;
  }

  // Separate into debtors (negative balance) and creditors (positive balance)
  const debtors: { person: string; amount: number }[] = [];
  const creditors: { person: string; amount: number }[] = [];

  for (const [person, balance] of Object.entries(balances)) {
    if (balance < -0.01) {
      debtors.push({ person, amount: -balance }); // amount owed
    } else if (balance > 0.01) {
      creditors.push({ person, amount: balance }); // amount to receive
    }
  }

  // Sort by amount descending (Greedy approach)
  debtors.sort((a, b) => b.amount - a.amount);
  creditors.sort((a, b) => b.amount - a.amount);

  const simplifiedEdges: DebtEdge[] = [];
  let d = 0; // debtor index
  let c = 0; // creditor index

  while (d < debtors.length && c < creditors.length) {
    const debtor = debtors[d];
    const creditor = creditors[c];

    const settledAmount = Math.min(debtor.amount, creditor.amount);

    simplifiedEdges.push({
      from: debtor.person,
      to: creditor.person,
      amount: Math.round(settledAmount * 100) / 100, // round to 2 decimals
    });

    debtor.amount -= settledAmount;
    creditor.amount -= settledAmount;

    if (debtor.amount < 0.01) d++;
    if (creditor.amount < 0.01) c++;
  }

  return simplifiedEdges;
}
